drivers/clock_control/clock_control_nrf2_hsfll.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_hsfll
+
+#include "clock_control_nrf2_common.h"
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* TODO: add support for other HSFLLs */
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+#include <ld_dvfs_handler.h>
+
+#define UPDATE_TIMEOUT_US K_USEC(1000)
+
+/* TODO: move to devicetree */
+static const uint32_t oppoint_freq[] = { MHZ(64), MHZ(128), MHZ(320) };
+static const uint32_t max_frequency = MHZ(320);
+
+struct hsfll_dev_data {
+	STRUCT_CLOCK_CONFIG(hsfll, ARRAY_SIZE(oppoint_freq)) clk_cfg;
+	struct k_timer timer;
+};
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
+
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	struct hsfll_dev_data *dev_data = dev->data;
+
+	k_timer_stop(&dev_data->timer);
```

Comment:
What if this function gets called after the timer fires? Is it ok to call `clock_config_update_end()` multiple times for one `_begin()`?

Reply:
`clock_config_update_end()` checks at its beginning if an update is in progress and returns if not. But indeed, it did not clear `FLAG_UPDATE_IN_PROGRESS` at that point, so another entry could cause issues in further processing. This has been fixed.

Reply:
Could you add API docs for the common module describing design principles like this? It would be easier to follow the programmers intention while reading the code.

Reply:
Added.